File: frontend/src/types/deal.ts
export type DealStage =
  | 'Lead In'
  | 'Qualified'
  | 'Proposal'
  | 'Negotiation'
  | 'Closed Won'
  | 'Closed Lost'

export const DEAL_STAGES: DealStage[] = [
  'Lead In',
  'Qualified',
  'Proposal',
  'Negotiation',
  'Closed Won',
  'Closed Lost',
]

export const OPEN_STAGES: DealStage[] = [
  'Lead In',
  'Qualified',
  'Proposal',
  'Negotiation',
]

export interface UserRef {
  id: number
  name: string
  email: string
}

export interface AccountRef {
  id: number
  name: string
}

export interface ContactRef {
  id: number
  name: string
}

export interface Deal {
  id: number
  ref_id: string
  title: string
  value: string
  stage: DealStage
  expected_close_date: string
  is_overdue: boolean
  is_archived: boolean
  loss_reason: string | null
  owner: UserRef
  account: AccountRef
  contact: ContactRef | null
  created_at: string
  updated_at: string
}

export interface DealSummary {
  id: number
  ref_id: string
  title: string
  value: string
  stage: DealStage
  expected_close_date: string
  is_overdue: boolean
  owner: UserRef
  account: AccountRef
}

export interface DealListResponse {
  total: number
  page: number
  limit: number
  items: DealSummary[]
}

export interface DealCreateRequest {
  title: string
  value: string
  stage: DealStage
  expected_close_date: string
  owner_id: number
  account_id: number
  contact_id?: number | null
}

export interface DealUpdateRequest {
  title?: string
  value?: string
  expected_close_date?: string
  contact_id?: number | null
}

export interface StageChangeRequest {
  stage: DealStage
  loss_reason?: string | null
}

export interface StageChangeResponse {
  ref_id: string
  previous_stage: DealStage
  new_stage: DealStage
  is_overdue: boolean
  updated_at: string
}

export interface DealComment {
  id: number
  body: string
  author: UserRef
  created_at: string
}

export interface CommentListResponse {
  total: number
  page: number
  limit: number
  items: DealComment[]
}

export interface ActivityLog {
  id: number
  action_type: string
  note: string | null
  actor: UserRef | null
  created_at: string
}

export interface ActivityLogResponse {
  total: number
  items: ActivityLog[]
}

export interface StageTotal {
  stage: DealStage
  deal_count: number
  total_value: string
  probability: string
  weighted_value: string
}

export interface ClosedWonTotal {
  deal_count: number
  total_value: string
}

export interface ForecastResponse {
  period_start: string
  period_end: string
  open_pipeline: StageTotal[]
  closed_won: ClosedWonTotal
  total_weighted_forecast: string
}

export interface DealFilterParams {
  stage?: DealStage
  owner_id?: number
  account_id?: number
  is_overdue?: boolean
  close_date_from?: string
  close_date_to?: string
  page?: number
  limit?: number
}

File: backend/app/schemas/deal.py
"""Pydantic v2 schemas for the Sales Pipeline deal endpoints."""

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.deal import DealStage


class UserRef(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    name: str


class AccountRef(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    name: str
    is_archived: bool


class ContactRef(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    name: str


class DealCreate(BaseModel):
    title: str = Field(min_length=1)
    value: Decimal = Field(default=Decimal("0.00"), ge=Decimal("0.00"))
    stage: DealStage = DealStage.LEAD_IN
    expected_close_date: date
    owner_id: int
    account_id: int
    contact_id: int | None = None


class DealUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    title: str | None = Field(default=None, min_length=1)
    value: Decimal | None = Field(default=None, ge=Decimal("0.00"))
    expected_close_date: date | None = None
    owner_id: int | None = None
    account_id: int | None = None
    contact_id: int | None = None


class DealRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    ref_id: str
    title: str
    value: Decimal
    stage: DealStage
    expected_close_date: date
    loss_reason: str | None
    is_overdue: bool
    is_archived: bool
    owner: UserRef
    account: AccountRef | None
    contact: ContactRef | None
    created_at: datetime
    updated_at: datetime


class DealSummary(BaseModel):
    """Lightweight read model for list views — same fields as DealRead."""
    model_config = ConfigDict(from_attributes=True)
    id: int
    ref_id: str
    title: str
    value: Decimal
    stage: DealStage
    expected_close_date: date
    is_overdue: bool
    is_archived: bool
    owner: UserRef
    account: AccountRef | None
    contact: ContactRef | None
    created_at: datetime
    updated_at: datetime


class DealListResponse(BaseModel):
    total: int
    page: int
    limit: int
    items: list[DealSummary]


class DealFilterParams(BaseModel):
    model_config = ConfigDict(populate_by_name=True)
    stage: DealStage | None = None
    owner_id: int | None = Field(default=None, alias="owner")
    account_id: int | None = None
    is_overdue: bool | None = None
    page: int = Field(default=1, ge=1)
    limit: int = Field(default=25, ge=1, le=100)


class StageChangeRequest(BaseModel):
    stage: DealStage
    loss_reason: str | None = None


class StageChangeResponse(BaseModel):
    ref_id: str
    previous_stage: DealStage
    new_stage: DealStage
    is_overdue: bool
    updated_at: datetime
